core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala

```diff
@@ -0,0 +1,170 @@
+  }
+}
+
+object FallbackStorage extends Logging {
+  /** We use one block manager id as a place holder. */
+  val FALLBACK_BLOCK_MANAGER_ID: BlockManagerId = BlockManagerId("fallback", "remote", 7337)
+
+  def getFallbackStorage(conf: SparkConf): Option[FallbackStorage] = {
+    if (conf.get(STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined) {
+      Some(new FallbackStorage(conf))
+    } else {
+      None
+    }
+  }
+
+  /** Register the fallback block manager and its RPC endpoint. */
+  def registerBlockManagerIfNeeded(master: BlockManagerMaster, conf: SparkConf): Unit = {
+    if (conf.get(STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined) {
+      master.registerBlockManager(
+        FALLBACK_BLOCK_MANAGER_ID, Array.empty[String], 0, 0, new NoopRpcEndpointRef(conf))
+    }
+  }
+
+  /** Report block status to block manager master and map output tracker master. */
+  private def reportBlockStatus(blockManager: BlockManager, blockId: BlockId, dataLength: Long) = {
+    assert(blockManager.master != null)
+    blockManager.master.updateBlockInfo(
+      FALLBACK_BLOCK_MANAGER_ID, blockId, StorageLevel.DISK_ONLY, memSize = 0, dataLength)
+  }
+
+  /**
+   * Read a ManagedBuffer.
+   */
+  def read(conf: SparkConf, blockId: BlockId): ManagedBuffer = {
+    logInfo(s"Read $blockId")
+    val fallbackPath = new Path(conf.get(STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).get)
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    val fallbackFileSystem = FileSystem.get(fallbackPath.toUri, hadoopConf)
+
+    val (shuffleId, mapId, startReduceId, endReduceId) = blockId match {
+      case id: ShuffleBlockId =>
+        (id.shuffleId, id.mapId, id.reduceId, id.reduceId + 1)
+      case batchId: ShuffleBlockBatchId =>
+        (batchId.shuffleId, batchId.mapId, batchId.startReduceId, batchId.endReduceId)
+      case _ =>
+        throw new IllegalArgumentException("unexpected shuffle block id format: " + blockId)
+    }
+
+    val name = ShuffleIndexBlockId(shuffleId, mapId, NOOP_REDUCE_ID).name
+    val indexFile = new Path(fallbackPath, s"$shuffleId/$name")
+    val start = startReduceId * 8L
+    val end = endReduceId * 8L
+    Utils.tryWithResource(fallbackFileSystem.open(indexFile)) { inputStream =>
+      Utils.tryWithResource(new DataInputStream(inputStream)) { index =>
+        index.skip(start)
+        val offset = index.readLong()
+        index.skip(end - (start + 8L))
+        val nextOffset = index.readLong()
+        val name = ShuffleDataBlockId(shuffleId, mapId, NOOP_REDUCE_ID).name
+        val dataFile = new Path(fallbackPath, s"$shuffleId/$name")
+        val f = fallbackFileSystem.open(dataFile)
+        val size = nextOffset - 1 - offset
+        logDebug(s"To byte array $size")
+        val array = new Array[Byte](size.toInt)
+        f.seek(offset)
+        f.read(array)
+        val startTimeNs = System.nanoTime()
+        logDebug(s"Took ${(System.nanoTime() - startTimeNs) / (1000 * 1000)}ms")
```

Comment:
Do you want to put `f.read` between `startTimeNs` and `logDebug`?

Reply:
Oops. Thanks.